stdlib/buffer.ml

```diff
@@ -197,6 +197,15 @@ let add_channel b ic len =
   if b.position + len > b.length then resize b len;
   add_channel_rec b ic len
 
+let add_channel_full b ic =
+  let rec loop b ic already_read =
+    let len = max 80 (b.length - b.position) in
+    let n = input_up_to ic b.buffer ~pos:b.position ~len in
+    let already_read = already_read + n in
+    if n < len then n
+    else loop b ic already_read
+  in loop b ic 0
```

Comment:
The `80` caught my eye, but something else cannot be right - there's no call to `resize`?

I got lost in the various issues leading to this PR - what's wrong with using `input` and just accepting whatever it gives back for the loop? It seems strange to be using one function `input_up_to` which loops over `input` to get some input and then putting another loop around that to keep reading?

Reply:
You are very correct, various things in this implementation were wrong (`then n` was also wrong). Apologies for the confusion; it looks like I've gone back-and-forth between a bit too many different choices to think clearly about it yesterday evening.

I have done another iteration that should fix the implementation. I'm also planning to write tests to exercise these functions, but I was sort of hoping that the discussion on "do we want these functions?" and "what names should they have?" would sort of settle down before we start tearing the implementations apart.

> what's wrong with using `input` and just accepting whatever it gives back for the loop

My initial idea was to move the "loop on partial input" logic from the Buffer module to the Stdlib module (and have other potential consumers benefit from it). As you suggest, we have to loop anyway so it may be just as easy to use `input`. I'll give it a try.

Reply:
@gasche Ooh, a general `loop on partial input` utility function in the stdlib sounds like an interesting idea.